src/go/rpk/pkg/serde/jsonschema.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+package serde
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/santhosh-tekuri/jsonschema/v6"
+	"github.com/twmb/franz-go/pkg/sr"
+)
+
+func compileJsonschema(ctx context.Context, cl *sr.Client, schema *sr.Schema) (*jsonschema.Schema, error) {
+	c := jsonschema.NewCompiler()
+	err := parseJsonschemaReferences(ctx, cl, schema, c)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse json schema references: %v", err)
+	}
+	sch, err := jsonschema.UnmarshalJSON(strings.NewReader(schema.Schema))
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal base schema: %v", err)
+	}
+	baseFileName := "redpanda_jsonschema.json"
+	err = c.AddResource(baseFileName, sch)
+	if err != nil {
+		return nil, fmt.Errorf("unable to add base json schema to the compiler resource: %v", err)
+	}
+	return c.Compile(baseFileName)
+}
+
+func newJsonschemaEncoder(schemaID int, jSchema *jsonschema.Schema) (serdeFunc, error) {
+	return func(record []byte) ([]byte, error) {
+		sch, err := jsonschema.UnmarshalJSON(bytes.NewReader(record))
+		if err != nil {
+			return nil, fmt.Errorf("unable to unmarshal record: %v", err)
+		}
+		err = jSchema.Validate(sch)
+		if err != nil {
+			return nil, fmt.Errorf("unable to validate json schema: %v", err)
+		}
+		// Append the magic byte + the schema ID bytes.
+		var serdeHeader sr.ConfluentHeader
+		h, err := serdeHeader.AppendEncode(nil, schemaID, nil)
+		if err != nil {
+			return nil, fmt.Errorf("unable to append header: %v", err)
+		}
+		return append(h, record...), nil
```

Comment:
Not worth it in this PR, but I wonder if we can have a fast follow up commit that does this prefix appending in a wrapper function in pkg/serde/serde.go